RecoLocalMuon/GEMSegment/plugins/ME0SegmentBuilder.cc

```diff
@@ -6,89 +6,75 @@
-    
-    std::vector<const ME0RecHit*> me0RecHits;
-    std::map<uint32_t,const ME0EtaPartition* > ens;
-    
-    // all detIds have been assigned to the to chamber
-    const ME0Chamber* chamber = geom_->chamber(enIt->first);    
-    for(auto rechit = enIt->second.begin(); rechit != enIt->second.end(); ++rechit) {
-      me0RecHits.push_back(*rechit);
-      ens[(*rechit)->me0Id()]=geom_->etaPartition((*rechit)->me0Id());
-    }    
-    ME0SegmentAlgorithmBase::ME0Ensemble ensemble(std::pair<const ME0Chamber*, std::map<uint32_t,const ME0EtaPartition*> >(chamber,ens));
-    
-    ME0DetId mid(enIt->first);
-    #ifdef EDM_ML_DEBUG
-    LogDebug("ME0SegmentBuilder") << "found " << me0RecHits.size() << " rechits in etapart " << mid;
-    #endif
-    
-    // given the chamber select the appropriate algo... and run it
-    std::vector<ME0Segment> segv = algo->run(ensemble, me0RecHits);
-    
-    #ifdef EDM_ML_DEBUG
-    LogDebug("ME0SegmentBuilder") << "found " << segv.size() << " segments in etapart " << mid;
-    #endif
-    
-    // Add the segments to master collection
-    // oc.put(mid, segv.begin(), segv.end());
-
-    // Add the segments to the chamber segment collection
-    // segment is defined from first partition of first layer    
-    ME0DetId midch = mid.chamberId();
-    ensembleSeg[midch.rawId()].insert(ensembleSeg[midch.rawId()].end(), segv.begin(), segv.end());
-  }
 
-  for(auto segIt=ensembleSeg.begin(); segIt != ensembleSeg.end(); ++segIt) {
-    // Add the segments to master collection
-    ME0DetId midch(segIt->first);
-    oc.put(midch, segIt->second.begin(), segIt->second.end());
+  std::map<ME0DetId, bool> foundChambers;
+  for(ME0RecHitCollection::const_iterator it = recHits->begin(); it != recHits->end(); it++) {
+	  const auto chId = it->me0Id().chamberId();
+	  auto chIt = foundChambers.find(chId);
+	  if(chIt !=  foundChambers.end()) continue;
+	  foundChambers[chId] = true;
+      ME0SegmentAlgorithmBase::HitAndPositionContainer hitAndPositions;
+      const ME0Chamber* chamber = geom_->chamber(chId);
+	  for(ME0RecHitCollection::const_iterator it2 = it; it2 != recHits->end(); it2++) {
+		  if(it2->me0Id().chamberId() != chId) continue;
+
+          const auto*  part = geom_->etaPartition(it2->me0Id());
+          GlobalPoint glb = part->toGlobal(it2->localPosition());
+          LocalPoint nLoc = chamber->toLocal(glb);
+          hitAndPositions.emplace_back(&(*it2),nLoc,glb,hitAndPositions.size());
+	  }
+
+      LogDebug("ME0Segment|ME0") << "found " << hitAndPositions.size() << " rechits in chamber " << *chIt;
+      //sort by layer
+        auto getLayer =[&](int iL) ->const ME0Layer* { //function is broken in the geo currently
+      	  for (auto layer : chamber->layers()){
+      		  if (layer->id().layer()==iL)
+      			  return layer;
+      	  }
+      	  return 0;
+        };
+        float z1 = getLayer(1)->position().z();
+        float z6 = getLayer(6)->position().z();
+        if(z1 < z6)
```

Comment:
That was a bug, the second case should have resulted in an opposite sorting order.